src/hardware/mouse/mouse_manymouse.cpp

```diff
@@ -0,0 +1,481 @@
+	}
+
+	Rescan();
+}
+
+void ManyMouseGlue::ShutdownIfSafe()
+{
+	if (is_mapping_in_effect || config_api_counter)
+		return;
+
+	ShutdownForced();
+}
+
+void ManyMouseGlue::ShutdownForced()
+{
+	if (!initialized)
+		return;
+
+	PIC_RemoveEvents(manymouse_tick);
+	ManyMouse_Quit();
+	ClearPhysicalMice();
+	num_mice    = 0;
+	initialized = false;
+}
+
+void ManyMouseGlue::StartConfigAPI()
+{
+	// Config API object is being created
+	++config_api_counter;
+	assert(config_api_counter > 0);
+}
+
+void ManyMouseGlue::StopConfigAPI()
+{
+	// Config API object is being destroyed
+	assert(config_api_counter > 0);
+	config_api_counter--;
+	ShutdownIfSafe();
+	if (!config_api_counter)
+		rescan_blocked_config = false;
+}
+
+void ManyMouseGlue::ClearPhysicalMice()
+{
+	mouse_info.physical.clear();
+	physical_devices.clear();
+	rel_x.clear();
+	rel_y.clear();
+}
+
+void ManyMouseGlue::Rescan()
+{
+	if (config_api_counter)
+		// Do not allow another rescan until MouseConfigAPI
+		// stops being used, it would be unsafe due to possible
+		// changes of physical device list size/indices
+		rescan_blocked_config = true;
+
+	ClearPhysicalMice();
+
+	for (uint8_t idx = 0; idx < num_mice; idx++) {
+		const auto name_utf8 = ManyMouse_DeviceName(idx);
+		std::string name;
+		UTF8_RenderForDos(name_utf8, name);
+
+		const char character_nbsp  = 0x7f; // non-breaking space
+		const char character_space = 0x20;
+
+		for (auto pos = name.size(); pos > 0; pos--) {
+			// Replace non-breaking space with a regular space
+			if (name[pos - 1] == character_nbsp)
+				name[pos - 1] = character_space;
+			// Remove non-ASCII and control characters
+			if (name[pos - 1] < character_space ||
+			    name[pos - 1] >= character_nbsp)
+				name.erase(pos - 1, 1);
+		}
+
+		// Try to rework into something useful names in the forms
+		// 'FooBar Corp FooBar Corp Incredible Mouse'
+		size_t pos = name.size() / 2 + 1;
+		while (--pos > 2) {
+			if (name[pos - 1] != ' ')
+				continue;
+			const std::string tmp = name.substr(0, pos);
+			if (name.rfind(tmp + tmp, 0) == std::string::npos)
+				continue;
+			name = name.substr(pos);
+			break;
+		}
+
+		// ManyMouse should limit device names to 64 characters,
+		// but make sure name is indeed limited in length, and
+		// strip trailing spaces
+		name.resize(std::min(static_cast<size_t>(64), name.size()));
+		while (!name.empty() && name.back() == ' ')
+			name.pop_back();
+
+		physical_devices.emplace_back(MousePhysical(name));
+		mouse_info.physical.emplace_back(MousePhysicalInfoEntry(
+		        static_cast<uint8_t>(physical_devices.size() - 1)));
+	}
+}
+
+void ManyMouseGlue::RescanIfSafe()
+{
+	if (rescan_blocked_config)
+		return;
+
+	ShutdownIfSafe();
+	InitIfNeeded();
+}
+
+bool ManyMouseGlue::ProbeForMapping(uint8_t &device_id)
+{
+	// Wait a little to speedup screen update
+	const auto pic_ticks_start = PIC_Ticks;
+	while (PIC_Ticks >= pic_ticks_start && PIC_Ticks - pic_ticks_start < 50)
```

Comment:
What's 50 represent? (move to constexpr)

Reply:
This is a time to wait idle in PIC ticks. Moved to `constexpr` and added a comment.